build/media_source/com_templates/vue/components/app.vue

```diff
@@ -0,0 +1,141 @@
+
+			<!-- Element -->
+			<draggable v-model="elementArray" handle=".handle">
+				<item v-for="element in elementArray" :key="element.key" :class="['row-wrapper']"
+						:item="element" :handleRequired="true"
+						@delete="deleteElement({ element })"
+						@edit="editElement({ element })"></item>
+			</draggable>
+			<!-- Element Ends -->
+
+			<button @click="addElement()" class="btn btn-success btn-block" type="button">
+				<span class="icon-new"></span>
+				<span>{{ translate('COM_TEMPLATES_ADD_ELEMENT') }}</span>
+			</button>
+
+			<!-- Modals -->
+			<add-element-modal id="add-element"></add-element-modal>
+		</div>
+	</div>
+</template>
+
+<script>
+  import {mapMutations, mapState} from 'vuex';
+  import draggable from 'vuedraggable';
+
+  export default {
+    computed: {
+      ...mapState([
+        'activeDevice',
+        'resolution',
+        'selectedSettings',
+      ]),
+      deviceWidth() {
+        return this.resolution[this.activeDevice];
+      },
+      elementArray: {
+        get() {
+          return this.$store.state.elementArray;
+        },
+        set(value) {
+          this.updateElementArray(value);
+        }
+      },
+    },
+    watch: {
+      elementArray: {
+        handler(newVal) {
+          document.getElementById('jform_params_grid').value = JSON.stringify(newVal);
+        },
+        deep: true,
+      },
+    },
+    created() {
+      this.mapElements(JSON.parse(document.getElementById('jform_params_grid').value));
+      this.checkAllowedElements();
+    },
+    mounted() {
+      if(document.getElementsByClassName('drag_component').length) {
+        var element = document.getElementsByClassName('drag_component')[0];
+        element.appendChild(document.getElementById('drag_component'));
+      }
+      if(document.getElementsByClassName('drag_message').length) {
+        var element = document.getElementsByClassName('drag_message')[0];
+        element.appendChild(document.getElementById('drag_message'));
+      }
+    },
+    methods: {
+      ...mapMutations([
+        'checkAllowedElements',
+        'mapElements',
+        'closeNav',
+        'deleteElement',
+        'setParent',
+        'updateElementArray',
+        'editElement',
+        'updateGrid',
+      ]),
+      addElement() {
+        this.setParent(this.elementArray);
+        this.$modal.show('add-element');
+      },
+      drag(event) {
+        event.dataTransfer.setData('text', event.target.id);
+      },
+      restorePosition(event, location) {
+        var element = document.getElementsByClassName('drag_' + location)[0];
+        if(location == 'component') {
+          element.__vue__.$data.element.options.component = false;
+        }
+        else {
+          element.__vue__.$data.element.options.message = false;
+        }
+        element.classList.remove('drag_' + location);
+        document.getElementById('placeholder_' + location).appendChild(document.getElementById('drag_' + location));
```

Comment:
can we use `$refs` ?

Reply:
I think it will result in code duplication as we cant use something like `'placeholder_' + location` with $refs, we would have to write it twice `placeholder_component` & `placeholder_message`